relay-server/src/actors/project.rs

```diff
@@ -379,6 +382,39 @@ enum GetOrFetch<'a> {
     Scheduled(&'a mut StateChannel),
 }
 
+/// Represents either the project state or an aggregation of metrics.
+///
+/// We have to delay rate limiting on metrics until we have a valid project state,
+/// So when we don't have one yet, we hold them in this aggregator until the project state arrives.
+///
+/// TODO: spool queued metrics to disk when the in-memory aggregator becomes too full.
+#[derive(Debug)]
+enum State {
+    Cached(Arc<ProjectState>),
+    Pending(Box<aggregator::Aggregator>),
+}
+
+impl State {
+    fn state_value(&self) -> Option<Arc<ProjectState>> {
+        match self {
+            State::Cached(state) => Some(Arc::clone(state)),
+            State::Pending(_) => None,
+        }
+    }
+
+    /// Sets the new `ProjectState`. If the variant was pending, the buckets will be returned.
+    fn set_state(&mut self, state: Arc<ProjectState>) -> Option<Vec<Bucket>> {
+        match std::mem::replace(self, Self::Cached(state)) {
+            State::Pending(agg) => Some(agg.into_buckets()),
+            State::Cached(_) => None,
+        }
+    }
+
+    fn new(config: AggregatorConfig) -> Self {
+        Self::Pending(Box::new(aggregator::Aggregator::new(config)))
```

Comment:
The aggregator created here should get a name, e.g. `"pre"`. Creating multiple aggregators with name `default` (through `Aggregator::new`) is undesirable because the name is used to tag statsd metrics that we collect, and we don't want to mix those for multiple aggregators.

Let's pass the name as an argument to `Aggregator::new`.

Reply:
oh yeah that makes sense!